frontends/p4/typeChecking/typeCheckDecl.cpp

```diff
@@ -314,4 +371,104 @@ const IR::Node *TypeInference::postorder(IR::Method *method) {
     return method;
 }
 
+/// get size int bits required to represent given constant
+static int getConstantsRepresentationSize(big_int val, bool isSigned) {
+    if (val < 0) {
+        val = -val;
+    }
+    int cnt = 0;
+    while (val > 0) {
+        ++cnt;
+        val >>= 1;
+    }
+    return cnt + int(isSigned);
+}
+
+/// Check if the value initializer fits into the underlying enum type. Emits error and returns false
+/// if it does not fit. Returns true if it fits.
+static bool checkEnumValueInitializer(const IR::Type_Bits *type, const IR::Expression *initializer,
+                                      const IR::Type_SerEnum *serEnum,
+                                      const IR::SerEnumMember *member) {
+    // validate the constant fits -- non-fitting enum constants should produce error
+    if (const auto *constant = initializer->to<IR::Constant>()) {
+        // signed values are two's complement, so [-2^(n-1)..2^(n-1)-1]
+        big_int low = type->isSigned ? -(big_int(1) << type->size - 1) : big_int(0);
+        big_int high = (big_int(1) << (type->isSigned ? type->size - 1 : type->size)) - 1;
+
+        if (constant->value < low || constant->value > high) {
+            int required = getConstantsRepresentationSize(constant->value, type->isSigned);
+            std::string extraMsg;
+            if (!type->isSigned && constant->value < low) {
+                extraMsg =
+                    str(boost::format(
+                            "the value %1% is negative, but the underlying type %2% is unsigned") %
+                        constant->value % type->toString());
+            } else {
+                extraMsg =
+                    str(boost::format("the value %1% requires %2% bits but the underlying "
+                                      "%3% type %4% only contains %5% bits") %
+                        constant->value % required % (type->isSigned ? "signed" : "unsigned") %
+                        type->toString() % type->size);
+            }
+            ::error(ErrorType::ERR_TYPE_ERROR,
+                    "%1%: Serialized enum constant value %2% is out of bounds of the underlying "
+                    "type %3%; %4%",
+                    member, constant->value, serEnum->type, extraMsg);
+            return false;
+        }
+    }
+    return true;
+}
+
+const IR::Node *TypeInferenceBase::postorder(const IR::SerEnumMember *member) {
+    /*
+      The type of the member is initially set in the Type_SerEnum preorder visitor.
+      Here we check additional constraints and we may correct the member.
+      if (done()) return member;
+    */
```

Comment:
This seems inconsistent with other comment styles in the file -- should it be double-slash?

Reply:
Yes, it should be. However, this was the original comment. I just moved the function to a different file.